### PR TITLE
[API][Product] Remove custom latest products endpoint

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | associated with https://github.com/Sylius/Sylius/pull/11402
| License         | MIT

* custom `/products/latest` endpoint has been removed
* `/products` endpoint could be used in the shop as well, if the user is logged in as an admin, they have the entire list of products provided, otherwise (Customer or Visitor) they have only products from current channel provided

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        return is_a($resourceClass, ProductInterface::class, true);
+    }
+
+    public function getCollection(string $resourceClass, string $operationName = null, array $context = []): iterable
+    {
+        $queryBuilder = $this->provideQueryBuilder($context);
+        $queryNameGenerator = new QueryNameGenerator();
+        foreach ($this->collectionExtensions as $extension) {
+            $extension->applyToCollection($queryBuilder, $queryNameGenerator, $resourceClass, $operationName, $context);
```
Comment (pamil) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php:63]: Why not register a custom collection extension to perform the logic implemented in `provideQueryBuilder` (add restriction if not logged in as admin)? This way there's no need to register a custom DataProvider.

Comment (GSadee) [reply src/Sylius/Bundle/ApiBundle/DataProvider/ProductCollectionDataProvider.php]: I've added a custom collection extension for products 👌

File: src/Sylius/Component/Core/Repository/ProductRepositoryInterface.php [hunk 2]
```diff
@@ -34,6 +34,8 @@ public function createShopListQueryBuilder(
         bool $includeAllDescendants = false
     ): QueryBuilder;
 
+    public function createListQueryBuilderByChannelAndLocaleCode(ChannelInterface $channel, string $localeCode): QueryBuilder;
```
Comment (pamil) [inline src/Sylius/Component/Core/Repository/ProductRepositoryInterface.php:37]: Is it still needed?

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByChannelAndLocaleCodeExtension.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        string $resourceClass,
+        string $operationName = null,
+        array $context = []
+    ): void {
+        if (!is_a($resourceClass, ProductInterface::class, true)) {
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles())) {
```
Comment (pamil) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ProductsByChannelAndLocaleCodeExtension.php:46]: Strict comparisons for the win!

```suggestion
        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
```

**Comment by lchrusciel** [comment]
Thanks, Grzegorz! :tada:

**Pull request merged**